backends/ubpf/target.cpp

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2019 Orange
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "target.h"
+
+namespace UBPF {
+
+    void UbpfTarget::emitIncludes(Util::SourceCodeBuilder *builder) const {
+        builder->append(
+                "#include <stdint.h>\n"
+                "#include <stdbool.h>\n"
+                "#include <stddef.h>\n"
+                "#include \"ubpf_common.h\"\n"
+                "\n");
+    }
+
+    void UbpfTarget::emitMain(Util::SourceCodeBuilder *builder,
+                              cstring functionName,
+                              cstring argName,
+                              cstring pktLen) const {
+        builder->appendFormat("uint64_t %s(void *%s, uint64_t %s)",
+                              functionName.c_str(), argName.c_str(), pktLen.c_str());
+    }
+
+    void UbpfTarget::emitTableLookup(Util::SourceCodeBuilder *builder,
+                                     cstring tblName,
+                                     cstring key,
+                                     UNUSED cstring value) const {
+        builder->appendFormat("ubpf_map_lookup(&%s, &%s)",
+                              tblName.c_str(), key.c_str());
+    }
+
+    void UbpfTarget::emitTableUpdate(Util::SourceCodeBuilder *builder,
+                                     cstring tblName,
+                                     cstring key,
+                                     cstring value) const {
+        builder->appendFormat("ubpf_map_update(&%s, &%s, %s)",
+                              tblName.c_str(), key.c_str(), value.c_str());
+    }
+
+    void UbpfTarget::emitGetPacketData(Util::SourceCodeBuilder *builder,
+                                       cstring ctxVar) const {
+        builder->appendFormat("ubpf_packet_data(%s)", ctxVar.c_str());
+    }
+
+    void UbpfTarget::emitUbpfHelpers(EBPF::CodeBuilder *builder) const {
+        builder->append(
```

Comment:
this looks a bit brittle. Is there a guarantee that the numbers associated with these functions will never change?

Reply:
This is a strict requirement, which is specific to target that we use (P4rt-OVS). Perhaps, it could be generalized in the future, but at the moment the BPF program would not work otherwise.